Standards/scs-0213-v1-k8s-nodes-anti-affinity.md

```diff
@@ -0,0 +1,123 @@
+But they also require more work during the setup of a service architecture, since nodes
+and pods need to be labelled and described consistently.
+
+In the case of SCS, affinity of the workloads themselves is not relevant, since this
+feature is mostly used by the customers of the providers.
+Instead, the expected standard should enable the Kubernetes cluster to handle Anti Affinity
+rules with a real physical separation as well as distributing the control plane over
+multiple machines in order to provide fault-tolerance during system outages.
+If the control plane survives an outage, a Kubernetes cluster can recover later on.
+
+## Design considerations
+
+SCS plans to require a Hard Anti Affinity and/or Redundancy for the control plane and
+a Soft Anti Affinity for workers. This means, that Anti Affinity would be required for
+the control planes and their pods and only optional (but encouraged) for workers.
+
+In order to achieve the goals for these components, meaning availability and fault tolerance
+for the control plane, in order to achieve an outage resistant cluster, and the availability
+promise given with Anti Affinity for pods on the worker nodes, a separation of nodes
+on the hardware or hypervisor level would need to be achieved. Since running multiple
+hypervisors on the same machine is in most cases not feasible and this approach
+wouldn't change availability for outages of individual machines, this should not be recommended.
+
+For the control plane, a reference to the "Kubernetes High Availability" [1]
+can be useful, since it provides two ways to set up a highly available cluster.
+Both approaches are very similar. The "Stacked Control Plane" requires less infrastructure,
+but also runs the risk of failed coupling, where if one node fails, the redundancy could be
+compromised due to the loss of a complete control plane instance.
+The "External ETCD" solves this problem, but also requires double the infrastructure, due
+to the externally incorporated etcd clusters.
+
+This also shows, that the wording "anti affinity" as used with Kubernetes pods is probably
+slightly misleading in the context of a Kubernetes control plane. It may consist of multiple
+pods with individual tasks, but distributing them over different nodes through Anti Affinity
+would probably still cascade the whole control plane into a complete failure, if one of
+the used nodes goes down. It could be possible to replicate specific important pods and
+assign them to different nodes, but at this point, a redundant setup like presented in [1] could be used.
+So Anti Affinity in this context probably means more like distribution over multiple
+physical machines, which needs to be planned beforehand on the machine/server level.
+
+Therefore would it be preferred for the control plane to use a redundant setup, which
+is separated over different physical machines, meaning at least half of the control plane
+nodes runs on a different physical machine as the rest.
```

Comment:
KaaS reference implementation uses the ClusterAPI approach. ClusterAPI approach knows the concept of “failure domains”.  “failure domain” is used to control where to place the k8s nodes and physically distribute them. For a HA k8s cluster with three control plane nodes, we can distribute the nodes into three different availability zones, to survive one availability zone failure. 

Currently, Multi-AZ deployment is not SCS KaaS default but could be implemented see [docs](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/blob/main/doc/usage/multi-az-and-multi-cloud-environments.md). 

Consider to mention the above, as IMO it is related to the sentence above.

Reply:
Updated the section with your suggested text (but I changed it a little bit).